File: parallelism.js
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads')
const th = require('ut/thread')

async function doit() {
    /* check cpu parallelism
    while (true) { }
    //*/

    // await th.sleep2(1000)
    for (let i = 0; i < 2000000000; ++i) { }
    console.log(workerData)
    parentPort.postMessage(null)
}


function main() {
    if (isMainThread) {
        var threads=5
        let count = 0
        let listener = async function (msg) {
            ++count
            for (let i = 0; i < 2000000000; ++i) { }
            // await th.sleep2(1000)
            console.log(count)
        }

        for (let i = 0; i < threads; ++i) {
            let w = new Worker(__filename, {workerData:'並行'});
            w.on('message', listener)
        }

    } else {
        doit()
    }
}
main()

File: parallelism.dart
import 'dart:io';
import 'dart:isolate';


main() async {
  var threads=5;
  var count = 0;
  var listener=() {
    ++count;
    sleep(Duration(seconds: 1));
    print(count);
  };

  for (var i = 0; i < threads; ++i) {
    var rp = new ReceivePort();
    var iso = await Isolate.spawn(doit, rp.sendPort);
    rp.listen((msg){
      listener();
      rp.close();
    });
  }
}

doit(SendPort sp) {
  /* check cpu parallelism
  while (true) {}
  // */
  sleep(Duration(seconds: 1));
  print('並行');
  sp.send(null);
}
